File: composeApp/src/androidMain/kotlin/co/jarias/flexapp/ui/screens/bingo/game_play/BingoGamePlayScreenViewModel.kt
package co.jarias.flexapp.ui.screens.bingo.game_play

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.jarias.flexapp.domain.usecase.CompleteGameUseCase
import co.jarias.flexapp.domain.usecase.GetGameStateUseCase
import co.jarias.flexapp.domain.usecase.MarkNumberUseCase
import co.jarias.flexapp.domain.usecase.RestartGameWithSameCardUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class BingoGamePlayScreenViewModel(
    private val getGameStateUseCase: GetGameStateUseCase,
    private val markNumberUseCase: MarkNumberUseCase,
    private val completeGameUseCase: CompleteGameUseCase,
    private val restartGameWithSameCardUseCase: RestartGameWithSameCardUseCase,
    private val gameId: Long = 0
) : ViewModel() {

    private val _state = MutableStateFlow(BingoGamePlayScreenState())
    val state: StateFlow<BingoGamePlayScreenState> = _state.asStateFlow()

    private var currentGameId: Long = gameId

    init {
        if (gameId != 0L) {
            loadGame(gameId)
        }
    }

    fun loadGame(gameId: Long) {
        currentGameId = gameId
        viewModelScope.launch {
            try {
                _state.value = _state.value.copy(isLoading = true, errorMessage = null)

                val gameState = getGameStateUseCase(gameId)

                if (gameState != null) {
                    _state.value = _state.value.copy(
                        gameState = gameState,
                        isLoading = false,
                        showWinDialog = gameState.isWon
                    )
                } else {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        errorMessage = "Game not found"
                    )
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    errorMessage = "Failed to load game: ${e.message}"
                )
            }
        }
    }

    fun onEvent(event: BingoGamePlayScreenEvents) {
        when (event) {
            is BingoGamePlayScreenEvents.OnNumberMarked -> markNumber(event.number)
            is BingoGamePlayScreenEvents.OnWinDialogDismissed -> dismissWinDialog()
            is BingoGamePlayScreenEvents.OnRetryClicked -> {
                loadGame(currentGameId)
            }
            is BingoGamePlayScreenEvents.OnPlayAgainClicked -> playAgain()
        }
    }

    private fun playAgain() {
        val currentState = _state.value
        val gameId = currentState.gameState?.game?.id ?: return
        
        viewModelScope.launch {
            try {
                _state.value = currentState.copy(isLoading = true)
                val newGame = restartGameWithSameCardUseCase(gameId)
                if (newGame != null) {
                    _state.value = currentState.copy(
                        isLoading = false,
                        navigateToNewGameId = newGame.id,
                        showWinDialog = false
                    )
                } else {
                    _state.value = currentState.copy(
                        isLoading = false,
                        errorMessage = "Failed to restart game"
                    )
                }
            } catch (e: Exception) {
                _state.value = currentState.copy(
                    isLoading = false,
                    errorMessage = "Failed to restart: ${e.message}"
                )
            }
        }
    }

    private fun markNumber(number: Int) {
        val currentState = _state.value
        val gameState = currentState.gameState
        val gameId = gameState?.game?.id ?: return

        viewModelScope.launch {
            try {
                val success = markNumberUseCase(gameId, number)

                if (success) {
                    val updatedGameState = getGameStateUseCase(gameId)

                    _state.value = currentState.copy(
                        gameState = updatedGameState,
                        lastMarkedNumber = number,
                        showWinDialog = updatedGameState?.isWon == true
                    )

                    if (updatedGameState?.isWon == true) {
                        completeGameUseCase(gameId)
                    }
                }
            } catch (e: Exception) {
                _state.value = currentState.copy(
                    errorMessage = "Failed to mark number: ${e.message}"
                )
            }
        }
    }

    private fun dismissWinDialog() {
        _state.value = _state.value.copy(showWinDialog = false)
    }
}

File: iosApp/iosApp/Screens/bingo/game_play/BingoGamePlayScreenViewModel.swift
import Foundation
import shared

class BingoGamePlayScreenViewModel: ObservableObject {
    @Published var state = BingoGamePlayScreenState()
    
    private let getGameStateUseCase: GetGameStateUseCase
    private let markNumberUseCase: MarkNumberUseCase
    private let completeGameUseCase: CompleteGameUseCase
    private let gameId: Int64
    
    init(
        getGameStateUseCase: GetGameStateUseCase,
        markNumberUseCase: MarkNumberUseCase,
        completeGameUseCase: CompleteGameUseCase,
        gameId: Int64
    ) {
        self.getGameStateUseCase = getGameStateUseCase
        self.markNumberUseCase = markNumberUseCase
        self.completeGameUseCase = completeGameUseCase
        self.gameId = gameId
        loadGame()
    }
    
    func loadGame() {
        state.isLoading = true
        state.errorMessage = nil
        
        getGameStateUseCase.invoke(gameId: gameId) { [weak self] gameState, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.state.isLoading = false
                if let state = gameState {
                    self.state.gameState = state
                    self.state.showWinDialog = state.isWon
                } else if let error = error {
                    self.state.errorMessage = "Failed to load game: \(error.localizedDescription)"
                } else {
                    self.state.errorMessage = "Game not found"
                }
            }
        }
    }
    
    func onEvent(_ event: BingoGamePlayScreenEvents) {
        switch event {
        case .onNumberMarked(let number):
            markNumber(number: number)
        case .onWinDialogDismissed:
            state.showWinDialog = false
        case .onRetryClicked:
            loadGame()
        }
    }
    
    private func markNumber(number: Int32) {
        markNumberUseCase.invoke(gameId: gameId, number: number) { [weak self] success, error in
            guard let self = self else { return }
            if success?.boolValue == true {
                self.getGameStateUseCase.invoke(gameId: self.gameId) { [weak self] updatedGameState, error in
                    DispatchQueue.main.async {
                        guard let self = self else { return }
                        self.state.gameState = updatedGameState
                        self.state.lastMarkedNumber = number
                        self.state.showWinDialog = updatedGameState?.isWon == true
                        
                        if updatedGameState?.isWon == true {
                            self.completeGameUseCase.invoke(gameId: self.gameId) { error in }
                        }
                    }
                }
            }
        }
    }
}
